File: src/main/c/jni-lib/src/events/QuoteMapping.cpp
// Copyright © 2023 Devexperts LLC. All rights reserved.
// This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
// If a copy of the MPL was not distributed with this file, You can obtain one at http://mozilla.org/MPL/2.0/.

#include "dxfeed/events/QuoteMapping.hpp"
#include "dxfeed/utils/ByteReader.hpp"
#include "dxfeed/utils/ByteWriter.hpp"

namespace dxfeed::jni {
  dxfg_quote_t* QuoteMapping::toQuote(ByteReader& reader) {
    auto* quote = new dxfg_quote_t();
    quote->market_event.event_type.clazz = DXFG_EVENT_QUOTE;
    quote->market_event.event_symbol = reader.readString();
    quote->market_event.event_time = reader.readLong();
    quote->time_millis_sequence = reader.readInt();
    quote->time_nano_part = reader.readInt();
    quote->bid_time = reader.readLong();
    quote->bid_exchange_code = reader.readInt16_t();
    quote->ask_time = reader.readLong();
    quote->ask_exchange_code = reader.readInt16_t();

    quote->bid_price = reader.readDouble();
    quote->bid_size = reader.readDouble();
    quote->ask_price = reader.readDouble();
    quote->ask_size = reader.readDouble();
    return quote;
  }

  void QuoteMapping::fromQuote(dxfg_quote_t* quote, ByteWriter& writer) {
    writer.writeString(quote->market_event.event_symbol);
    writer.writeInt64_t(quote->market_event.event_time);
    writer.writeInt32_t(quote->time_millis_sequence);
    writer.writeInt32_t(quote->time_nano_part);
    writer.writeInt64_t(quote->bid_time);
    writer.writeInt16_t(quote->bid_exchange_code);
    writer.writeInt64_t(quote->ask_time);
    writer.writeInt16_t(quote->ask_exchange_code);

    writer.writeDouble(quote->bid_price);
    writer.writeDouble(quote->bid_size);
    writer.writeDouble(quote->ask_price);
    writer.writeDouble(quote->ask_size);
  }
}

File: src/main/java/com/dxfeed/api/serializers/QuoteMapping.java
// Copyright © 2023 Devexperts LLC. All rights reserved.
// This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
// If a copy of the MPL was not distributed with this file, You can obtain one at http://mozilla.org/MPL/2.0/.

package com.dxfeed.api.serializers;

import com.dxfeed.api.buffers.ByteBuffer;
import com.dxfeed.api.buffers.DoubleBuffer;
import com.dxfeed.api.buffers.ByteReader;
import com.dxfeed.event.market.DxFeedEventMarketPackagePrivate;
import com.dxfeed.event.market.Quote;

public class QuoteMapping {
  /**
   * typedef struct dxfg_market_event_t {
   *    dxfg_event_type_t event_type;
   *    const char *event_symbol;
   *    int64_t event_time;
   * } dxfg_market_event_t;
   *
   * https://github.com/dxFeed/dxfeed-graal-native-sdk/blob/main/src/main/c/api/dxfg_events.h#L140
   *
   * typedef struct dxfg_quote_t {
   *    dxfg_market_event_t market_event;
   *    int32_t time_millis_sequence;
   *    int32_t time_nano_part;
   *    int64_t bid_time;
   *    int16_t bid_exchange_code;
   *    double bid_price;
   *    double bid_size;
   *    int64_t ask_time;
   *    int16_t ask_exchange_code;
   *    double ask_price;
   *    double ask_size;
   * } dxfg_quote_t;
   */
  public static void toNative(Quote event, ByteBuffer pBytes, DoubleBuffer pDoubles) {
    // BYTE DATA
    pBytes.writeString(event.getEventSymbol());                                     // 2 + eventSymbolLength
    pBytes.writeLong(event.getEventTime());                                         // 8
    pBytes.writeInt(DxFeedEventMarketPackagePrivate.getTimeMillisSequence(event));  // 4
    pBytes.writeInt(event.getTimeNanoPart());                                       // 4
    pBytes.writeLong(event.getBidTime());                                           // 8
    pBytes.writeChar(event.getBidExchangeCode());                                   // 2
    pBytes.writeLong(event.getAskTime());                                           // 8
    pBytes.writeChar(event.getAskExchangeCode());                                   // 2

    // DOUBLE DATA
    pDoubles.write(event.getBidPrice());
    pDoubles.write(event.getBidSizeAsDouble());
    pDoubles.write(event.getAskPrice());
    pDoubles.write(event.getAskSizeAsDouble());
  }

  public static Quote fromNative(ByteReader reader) {
    Quote quote = new Quote();
    quote.setEventSymbol(reader.readString());
    quote.setEventTime(reader.readLong());
    DxFeedEventMarketPackagePrivate.setTimeMillisSequence(quote, reader.readInt());
    quote.setTimeNanoPart(reader.readInt());
    quote.setBidTime(reader.readLong());
    quote.setBidExchangeCode(reader.readChar());
    quote.setAskTime(reader.readLong());
    quote.setAskExchangeCode(reader.readChar());

    quote.setBidPrice(reader.readDouble());
    quote.setBidSizeAsDouble(reader.readDouble());
    quote.setAskPrice(reader.readDouble());
    quote.setAskSizeAsDouble(reader.readDouble());
    return quote;
  }
}
